src/components/pages/devices/flyouts/deviceJobs/deviceJobTags.js

```diff
@@ -1,20 +1,304 @@
 // Copyright (c) Microsoft. All rights reserved.
 
 import React from 'react';
+import { Link } from "react-router-dom";
+import { Observable } from 'rxjs';
+import update from 'immutability-helper';
 
+import { IoTHubManagerService } from 'services';
+import { toSubmitTagsJobRequestModel } from 'services/models';
+import { LinkedComponent } from 'utilities';
+import { svgs, Validator } from 'utilities';
 import {
+  Btn,
+  BtnToolbar,
+  ErrorMsg,
+  FormControl,
+  FormGroup,
+  FormLabel,
   FormSection,
+  Indicator,
   SectionDesc,
-  SectionHeader
-} from '../../../../shared';
-
-export const DeviceJobTags = (props) => {
-  const { t, devices } = props;
-  return (
-    <FormSection className="device-job-tags-container">
-      <SectionHeader>{t('devices.flyouts.jobs.tags.title')}</SectionHeader>
-      <SectionDesc>{t('devices.flyouts.jobs.tags.description')}</SectionDesc>
-      <pre>{JSON.stringify(devices, null, 2)}</pre>
-    </FormSection>
-  )
+  SectionHeader,
+  SummaryCount,
+  SummarySection,
+  Svg
+} from 'components/shared';
+import {
+  PropertyGrid as Grid,
+  PropertyGridBody as GridBody,
+  PropertyGridHeader as GridHeader,
+  PropertyRow as Row,
+  PropertyCell as Cell
+} from 'components/pages/devices/flyouts/deviceDetails/propertyGrid';
+
+const isNumeric = value => typeof value === 'number' || !isNaN(parseInt(value, 10));
+const isAlphaNumericRegex = /^[a-zA-Z0-9]*$/;
+const nonAlphaNumeric = x => !x.match(isAlphaNumericRegex);
+
+const TagJobConstants = {
+  multipleValues: 'Multiple',
+  stringType: 'Text',
+  numberType: 'Number'
 };
+
+const initialState = {
+  isPending: false,
+  error: undefined,
+  successCount: 0,
+  changesApplied: false,
+  jobName: undefined,
+  jobId: undefined,
+  commonTags: [],
+  deletedTags: []
+};
+
+const newTag = () => ({
+  name: '',
+  value: '',
+  type: TagJobConstants.stringType
+});
+
+export class DeviceJobTags extends LinkedComponent {
+  constructor(props) {
+    super(props);
+    this.state = initialState;
+
+    // Linked components
+    this.jobNameLink = this.linkTo('jobName')
+      .reject(nonAlphaNumeric)
+      .check(Validator.notEmpty, () => this.props.t('devices.flyouts.jobs.validation.required'));
+
+    this.tagsLink = this.linkTo('commonTags');
+  }
+
+  componentDidMount() {
+    if (this.props.devices) {
+      this.populateState(this.props.devices);
+    }
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.devices && (this.props.devices || []).length !== nextProps.devices.length) {
+      this.populateState(nextProps.devices);
+    }
+  }
+
+  populateState(devices) {
+    // Create a stream of devices
+    const deviceStream = Observable.of(devices).flatMap(_ => _);
```

Comment:
I think this is a better approach. The current version is roughly n^2. This new version is roughly 2n.

I wrote this blind btw so might need some fixing up.

```
Observable.from(devices)
  .map(({ tags }) => new Set(Object.keys(tags)))
  .reduce((commonTags, deviceTags) => 
    commonTags
      ? new Set([...commonTags].filter(tag => deviceTags.has(tag)))
      : deviceTags
  ) // At this point, a stream of a single event. A common 
  .flatMap(commonTagsSet =>
    Observable.from(devices)
      .flatMap(({ tags }) => Object.entries(tags))
      .filter(([ tag ]) => commonTagsSet.has(tag))
  )
  .distinct(([ tagName, tagVal ]) => `${tagName} ${tagVal}`)
  .reduce((acc, [ tagName, tagVal ]) => update(acc, {
    [tagName]: { $autoArray: { 
      $push: [ tagVal ] 
    }}
  }), {})
  .flatMap(tagToValMap => Object.entries(tagToValMap))
  .reduce(
    (newState, [ name, values ]) => {
      const value = values.length === 1 ? values[0] : TagJobConstants.multipleValues;
      const type = values.every(isNumeric) ? TagJobConstants.numberType : TagJobConstants.stringType;
      return update(newState, {
        commonTags: { $push: [{ name, value, type }] }
      });
    },
    initialState
  ).subscribe(newState => this.setState(newState));
```